### DISCUSSION TITLE
How to change serialization of referenced pydantic models?

### DISCUSSION THREAD

**Discussion by SvenPVoigt** [pydantic/pydantic#2622] [Question]
I have a model that references other pydantic models, including itself. These models should serialize with some type of id instead of the entire object when they are referenced in another model. How would one accomplish this?

First, my problem setup. I have two pydantic models: a `phone` and a `user`, which can be references in a `user` model. When serializing the `user` model, I want to represent the `phone` and `user` references with the ids `number` and `SSN`, respectively. The serialization of the `user` reference might be more tricky because it requires updating forward references. So first, consider the case where the `user` model has custom serializations for a birthday attribute of type `datetime` and a phone attribute of type `phone`.

```python
from pydantic import BaseModel
from datetime import datetime

class Phone(BaseModel):
    manufacturer: str
    number: int

class User(BaseModel):
    name: str
    SSN: int
    birthday: datetime
    phone: Phone
    friend: "User"=None
        
    class Config:
        json_encoders = {
            datetime: lambda v: v.timestamp(),
            Phone: lambda v: v.number if v else None
        }
        
User.update_forward_refs()
```

As seen in the following example, the custom serialization of `datetime` works just as in the documentation. 

```python
iphone = Phone(manufacturer="Apple", number=18002752273)
galaxy = Phone(manufacturer="Samsung", number=18007267864)
        
timon = User(name="Timon", SSN=123, birthday=datetime(1993, 6, 1), phone=iphone)
pumbaa = User(name="Pumbaa", SSN=234, birthday=datetime(1993, 5, 15), phone=galaxy)

timon.friend = pumbaa

print( timon.json(indent=2) )
```

`Actual Output`
```json
{
  "name": "Timon",
  "SSN": 123,
  "birthday": 738907200.0,
  "friend": {
    "name": "Pumbaa",
    "SSN": 234,
    "birthday": 737438400.0,
    "friend": null,
    "phone": {
      "manufacturer": "Samsung",
      "number": 18007267864
    }
  },
  "phone": {
    "manufacturer": "Apple",
    "number": 18002752273
  }
}
```

However, the problem is that when a reference to a pydantic model is serialized, that it serializes like the `basemodel`. The documentation states that subclasses serialize like the parent class, but shouldn't a custom serialization for the subclass take preference? I would have expected the custom serialization function to yield `key` `value` pairs for phone like `phone: xxxxxxxx` (expected output shown below). However, the entire pydantic model is serialized as in the previous output.

`Expected Output`
```json
{
  "name": "Timon",
  "SSN": 123,
  "birthday": 738907200.0,
  "friend": {
    "name": "Pumbaa",
    "SSN": 234,
    "birthday": 737438400.0,
    "friend": null,
    "phone": 18007267864
  },
  "phone": 18002752273
}
```

Further, if there is a solution for the previous problem, how would I do forward references in custom serialization? I have two ideas, as shown below. First, I put the class name in `""` in the `json_decoders object`. However, I am not certain that pydantic will forward update this reference. The other idea is to update the `json_encoders` object after the `User` model has been defined. Would either of these work?

```python
class User(BaseModel):
    name: str
    SSN: int
    birthday: datetime
    phone: Phone
    friend: "User"=None
        
    class Config:
        json_encoders = {
            datetime: lambda v: v.timestamp(),
            Phone: lambda v: v.number if v else None,
            "User": lambda v: v.SSN
        }
        
User.update_forward_refs()

User.Config.json_encoders.update({
    User: lambda v: v.SSN
})
```

Expected output of serializing the timon object and using the above classes would be:

`Expected Output`
```json
{
  "name": "Timon",
  "SSN": 123,
  "birthday": 738907200.0,
  "friend": 234,
  "phone": 18002752273
}
```

**Comment by PrettyWood** [comment]
Hi @SvenPVoigt 
I totally agree with you. The issue is that we would always convert first to dict before serializing so it was impossible to customize how another pydantic model should be serialized.
I opened https://github.com/samuelcolvin/pydantic/pull/2650 and also added a way to customize "self-reference".
